keras/data/enqueuers.py

```diff
@@ -0,0 +1,248 @@
+import itertools
+import multiprocessing
+import random
+import threading
+import time
+from abc import abstractmethod
+from multiprocessing.pool import ThreadPool
+
+import numpy as np
+
+try:
+    import queue
+except ImportError:
+    import Queue as queue
+
+
+def get_index(ds, i):
+    """ Quick fix for Python2, otherwise, it cannot be pickled.
+
+    # Arguments
+        ds: a Dataset object
+        i: index
+
+    # Returns
+        The value at index `i`.
+    """
+    return ds[i]
+
+
+class DatasetHandler():
+    """Base class to enqueue datasets."""
+
+    @abstractmethod
+    def is_running(self):
+        raise NotImplemented
+
+    @abstractmethod
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, threads could block on put())
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def get(self):
```

Comment:
This is a limitation of Keras itself since fit_generator is calling sess.run with a feed_dict.
I don't think we want to go this deep :P Or maybe for a future PR?